src/destructors.md

```diff
@@ -37,57 +33,344 @@ impl Drop for ShowOnDrop {
     }
 }
 
+let mut overwritten = ShowOnDrop("drops when overwritten");
+overwritten = ShowOnDrop("drops when scope ends");
+
+let tuple = (ShowOnDrop("Tuple first"), ShowOnDrop("Tuple second"));
+
+let moved;
+// No destructor run on assignment.
+moved = ShowOnDrop("Drops when moved");
+// drops now, but is then uninitialized
+moved;
+let uninitialized: ShowOnDrop;
+// Only the first element is dropped
+let mut partially_initialized: (ShowOnDrop, ShowOnDrop);
+partially_initialized.0 = ShowOnDrop("Partial tuple first");
+```
+
+## Drop scopes
```

Comment:
Is all of this behavior specified/decided somewhere.. is some of it documenting current rustc behavior or are all parts decided upon by e.g. the lang team or e.g. by necessity due to NLL?

Reply:
It's a description of the current behaviour, which is mostly observable. That said, some of it could be changed: RFC 66 proposes this, but it is still unimplemented.

Reply:
OK; my interest is mainly in not additionally tying the language team's hand due to what we add here; if the text could be clarified to note what are stable guarantees and what are not that'd be helpful.

Reply:
The documentation is "in the code", as they predate RFCs. Thanks @matthewjasper for writing it up, I too had an "in progress" write-up somewhere (part of a larger amendment to RFC 66...) but I'm happy to have someone else take a crack at it.

Some relevant links:

* [terminating scopes](https://github.com/rust-lang/rust/blob/df768c5c8fcb361c4dc94b4c776d6a78c12862e1/src/librustc_passes/region.rs#L58-L77)
* [the rules for temporary scopes and lets](https://github.com/rust-lang/rust/blob/df768c5c8fcb361c4dc94b4c776d6a78c12862e1/src/librustc_passes/region.rs#L438-L492)